mlir/lib/Tools/PDLL/Parser/Parser.cpp

```diff
@@ -956,10 +980,24 @@ LogicalResult Parser::parseInclude(SmallVectorImpl<ast::Decl *> &decls) {
   // Check the type of include. If ending with `.pdll`, this is another pdl file
   // to be parsed along with the current module.
   if (filename.ends_with(".pdll")) {
-    if (failed(lexer.pushInclude(filename, fileLoc)))
+    std::string includedFile;
+    if (!lexer.getSourceMgr().OpenIncludeFile(filename.str(), includedFile))
       return emitError(fileLoc,
                        "unable to open include file `" + filename + "`");
 
+    // Check if included has been already processed
+    if (alreadyIncludedFiles.count(includedFile)) {
+      if (lexer.isLexingMainFile()) {
+        emitWarning(
```

Comment:
Why do we want this warning? Isn't that exactly what `#once` is used for? (I'm thinking that clang doesn't output a warning in this case either)

Reply:
I thought this would be helpful to keep includes cleaner. A potential case is shown here: https://github.com/Xilinx/llvm-project/pull/210/commits/2057276ddbb38d910f11b1a51ed402c7ba23379a#diff-a535c1080585a8dd9edda1501e7d9af5357c12060c51cd5cb2cac56e9a8a83ac

Second header is included already by first, so we would get a warning and we can remove it. Let me now if you still think this is overshooting

Reply:
Warning is now only emitted when running in lsp mode